tests/test_etherscan.py

```diff
@@ -1,21 +1,79 @@
 import pytest
 from ape import networks
+from ape.api.explorers import ExplorerAPI
 from requests import Response
 
+from ape_etherscan import NETWORKS
+
+ADDRESS = "https://etherscan.io/address/0xAb5801a7D398351b8bE11C439e05C5B3259aeC9B"
+TRANSACTION = "0x0da22730986e96aaaf5cedd5082fea9fd82269e41b0ee020d966aa9de491d2e6"
+
 
 @pytest.fixture
 def etherscan_abi_response(mocker):
     response = mocker.MagicMock(spec=Response)
-    f = open("tests/get_contract_response.json")
-    response.json.return_value = json.load(f)
-    f.close()
-    return response
+    test_data_path = Path(__file__).parent / "get_contract_response.json"
+    with open(test_data_path) as response_data_file:
+        response.json.return_value = json.load(response_data_file)
+        yield response
+
+
+def get_explorer(network_name: str) -> ExplorerAPI:
+    return getattr(networks.ethereum, network_name).explorer
+
+
+def setup_mock_get(mocker, etherscan_abi_response, expected_params):
+    get_patch = mocker.patch("ape_etherscan.client.requests")
+
+    def get(base_uri, params=None, *args, **kwargs):
+        # Request will fail if made with incorrect parameters.
+        assert base_uri == "https://api.etherscan.io/api"
+        assert params == expected_params
+        return etherscan_abi_response
+
+    get_patch.get.side_effect = get
+    return get_patch
+
+
+@pytest.mark.parametrize(
+    "network,expected_prefix,address",
+    [(NETWORKS[0], "etherscan.io", ADDRESS), (NETWORKS[1], "ropsten.etherscan.io", ADDRESS)],
+)
+def test_get_address_url(network, expected_prefix, address):
+    expected = f"https://{expected_prefix}/address/{ADDRESS}"
+    explorer = get_explorer(network)
+    actual = explorer.get_address_url(address)  # type: ignore
+    assert actual == expected
+
+
+@pytest.mark.parametrize(
+    "network,expected_prefix,tx_hash",
+    [
+        (NETWORKS[0], "etherscan.io", TRANSACTION),
+        (NETWORKS[1], "ropsten.etherscan.io", TRANSACTION),
+    ],
+)
+def test_get_transaction_url(network, expected_prefix, tx_hash):
+    expected = f"https://{expected_prefix}/tx/{tx_hash}"
+    explorer = get_explorer(network)
+    actual = explorer.get_transaction_url(tx_hash)
+    assert actual == expected
 
 
 def test_get_contract_type(mocker, etherscan_abi_response):
-    get_patch = mocker.patch("ape_etherscan.explorer.requests")
-    get_patch.get.return_value = etherscan_abi_response
-    actual = networks.ethereum.mainnet.explorer.get_contract_type("TEST")
-    assert actual.contractName == "BoredApeYachtClub"
+    expected_params = {
+        "module": "contract",
+        "action": "getsourcecode",
+        "address": ADDRESS,
+    }
+    setup_mock_get(mocker, etherscan_abi_response, expected_params)
+
+    explorer = get_explorer("mainnet")
+    actual = explorer.get_contract_type(ADDRESS)  # type: ignore
+
+    # Name comes from the 'get_contract_response.json' file
+    expected = "BoredApeYachtClub"
+    assert actual.contractName == expected
```

Comment:
This is the one that would work well with parameterization (maybe with integration test)

Try some contracts and check if you get a contract type name or nothing

But maybe this is all overkill, as you said in the other comments re: rate limiting and directly testing etherscan

Reply:
We are actually mocking the response here in the existing test.

I would like to make sure we handle these edge cases though. If the responses differ, we can still add tests making sure we handle those situations.

Do you have any good examples of proxy contracts e.g. stuff you like me to test here?

Reply:
example proxy: https://github.com/yearn/yearn-vaults/blob/main/contracts/Registry.vy#L164

This contract (`Registry`) collects new deployments of the vaults, and allows you to make EIP-1167 proxies to them.

Here's one in prod:
- `Registry`: https://etherscan.io/address/v2.registry.ychad.eth#code
- `Vault`: https://etherscan.io/address/0x9C13e225AE007731caA49Fd17A41379ab1a489F4#code
- EIP-1167 proxy to `Vault`: https://etherscan.io/address/0xa354F35829Ae975e850e23e9615b11Da1B3dC4DE#code